### DISCUSSION TITLE
@map("_id") for mongodb is useless

### DISCUSSION THREAD

**Discussion by jasonkuhrt** [prisma/prisma#13966] [Q&A]
## Problem

<img width="933" alt="CleanShot 2022-06-21 at 22 08 38@2x" src="https://user-images.githubusercontent.com/284476/174928112-64cfc699-a437-477d-96b6-2203bec295da.png">

## Suggested solution

The only valid thing to do is add `@map("_id")`. Even the error message  says this. Why can't the machine 1) do this for me and 2) Prisma abstract it (by default map it to `_id`).

## Alternatives

N/A

## Additional context

N/A

**Comment by janpio** [comment]
> Why can't the machine 1) do this for me

The machine you are using is a (in this case an implicit) validation command. It can not just change your Prisma schema file. It can only tell you what needs to be done somehow.

The other machine you are using is VSCode (I assume, as it tells you there are not quick fixes available) which might or might not have the Prisma extension installed (the syntax highlighting kinda indicates it is). For that we have a feature request to actually ad a close to "machine does it for you" solution: https://github.com/prisma/language-tools/issues/762 It will still need another click from you though.

If you have a better suggestion for another machine in play here that could automatically fix this without also confusing users by changing their schema, please let me know. I understand the pain here.

> Why can't ... 2) Prisma abstract it (by default map it to _id).

Because that is not how Prisma schema field definitions work - we have a name that needs to follow `[a-Z0-9]*` and a `map` attribute that can be used to define database side names that do not follow this regex. When the PSL syntax was invented, no one thought of MongoDB enforcing the naming of `_id` for PK fields. So now that clashes unfortunately.

**Reply by jasonkuhrt** [reply]
Thanks @janpio. But I still don't understand why Prisma internally, when Mongodb is the provider, imply the presence of attribute `map("_id")` for all `@id` fields. "Machine" here could be Prisma Client runtime, and/or Prisma Query Engine.

**Reply by janpio** [reply]
Prisma does not. It needs you to explicitly set it in the schema, so we do not add magic to the meaning of `@id` in MongoDB only. The way to redefine what a field should be called in Prisma is `map`.

(That this is needed of course a limitation caused by Prisma schema's limitation in not allowing leading underscores in field names - otherwise it could just be `_id @id @db.ObjectId`)

**Reply by jasonkuhrt** [reply]
I think there's sufficient overlap at this point between the underlying theme of this thread and https://github.com/prisma/prisma/discussions/13967 that I'll just focus over there, otherwise I'll probably be writing almost the same things twice right now.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
